src/AzureAuth/Ado/AdoToken.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Authentication.AzureAuth.Ado
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    using Microsoft.Office.Lasso.Interfaces;
+
+    /// <summary>
+    /// Business logic for either getting an ADO PAT from <see cref="IEnv"/>
+    /// or doing normal AAD Authentication.
+    /// </summary>
+    public static class AdoToken
+    {
+        private static readonly IEnumerable<string> PatEnvVars = new[]
+        {
+            EnvVars.AdoPat,
+            EnvVars.SystemAccessToken,
+        };
+
+        /// <summary>
+        /// Get either an ADO PAT from the environment. If one is not set, then do normal ADO AAD Auth.
+        /// </summary>
+        /// <param name="env">The <see cref="IEnv"/> to use for interacting with the environment.</param>
+        /// <returns>A PAT from the env if set, otherwise null.</returns>
+        public static PatResult PatFromEnv(IEnv env)
+        {
+            foreach (var envVar in PatEnvVars)
+            {
+                var pat = env.Get(envVar);
+                var exists = !string.IsNullOrEmpty(pat);
+
+                if (exists)
+                {
+                    return new ()
+                    {
+                        Exists = exists,
+                        Value = pat,
+                        EnvVarSource = envVar,
+                    };
+                }
+            }
+
+            return new ()
+            {
+                Exists = false,
+            };

```

Comment:
We only implement this if we came out of the foreach loop, which also means `Exists is false`. Any reason we set it to false again?

Reply:
I don't think `Exists` exists if we come out of the loop. A brand new `PatResult` is being created here. Are you suggesting we let a default be used? I think that might be `null` in this case. 🤔

Reply:
The default value of `bool` is false, but this is not obvious. Strings default to `null` and this is the `default` but it's clearer to set it directly, epecially with primitive types like bool and int.